free.1
.Dd March 10, 2008
.Dt FREE 1
.Os
.Sh NAME
.Nm free
.Nd Display the amount of free and used memory in the system
.Sh SYNOPSIS
.Nm
.Fl b
.Op Fl s Ar delay
.Nm
.Fl k
.Op Fl s Ar delay
.Nm
.Fl m
.Op Fl s Ar delay
.Nm
.Fl g
.Op Fl s Ar delay
.Nm
.Fl Vh?
.Sh DESCRIPTION
The
.Nm
utility displays the total amount of free and used physical memory in the
system, as well as the active, inactive, and wired page counts.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl b
Display memory sizes in bytes.
.It Fl k
Display memory sizes in kilobytes (default if no options given).
.It Fl m
Display memory sizes in megabytes.
.It Fl g
Display memory sizes in gigabytes.
.It Fl s Ar delay
Sleep for
.Ar delay
seconds before display memory statistics again.  This option causes free to
continuously poll the system until SIGINT is given.
.It Fl V
Display version information.
.El
.Pp
The
.Fl h
and
.Fl ?
options display usage information.
.Sh NOTES
.Pp
.Nm
is modeled after the free utility on Linux systems (the one from the procps
project).  This version aims to offer similar functionality, but due to
design differences between Linux and Darwin, that is not always possible.
.Pp
The columns that free displays may be a bit confusing.  The first one, total,
display the total amount of memory for that area.  The used column displays
the amount of total memory that is currently in use.  The free column displays
the amount of total memory that is currently in use.  The active column displays
the number of active pages in the used memory.  The inactive column shows the
number of inactive pages in the used memory.  And the wired column displays
the number of wired down pages in the used memory.
.Pp
The Mem: row uses these formulas:
.Dl total = used + free
.Dl used = active + inactive + wired
.Sh SEE ALSO
.Xr hostinfo 1 ,
.Xr ps 1 ,
.Xr top 1
.Sh BUGS
.Pp
Report bugs at http://code.google.com/p/darwin-free/
.Sh AUTHORS
.Pp
Written by David Cantrell <david.l.cantrell@gmail.com>
.Pp
Contributions from Logan Johnson <logan.johnson@gmail.com> and
Jim Bair <tseuhpsyde@gmail.com>


free.c
/*
 * free.c
 * Memory usage utility for Darwin & MacOS X (similar to 'free' under Linux)
 *
 * by:  David Cantrell <david.l.cantrell@gmail.com>
 * Copyright (C) 2002, David Cantrell, Atlanta, GA, USA.
 * Copyright (C) 2008, David Cantrell, Honolulu, HI, USA.
 *
 * Licensed under the GNU Lesser General Public License version 2.1 or later.
 * See COPYING.LIB for licensing details.
 */

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <libgen.h>
#include <mach/mach.h>
#include <mach/mach_error.h>
#include <mach/kern_return.h>
#include <mach/host_info.h>
#include <mach/host_priv.h>

#include "free.h"

extern char *optarg;
extern int optind;

static void set_units(int *units, int type) {
    if (*units != -1) {
        fprintf(stderr, COMBINED_UNIT_OPTIONS);
        fflush(stderr);
        exit(EXIT_FAILURE);
    }

    *units = type;
}

int main(int argc, char **argv) {
    int poll = 0, div = 1, units = -1;
    char c;
    kern_return_t ke = KERN_SUCCESS;
    mach_port_t host, task;
    vm_size_t hps;
    vm_statistics_data_t hs;
    struct host_basic_info hbi;
    mem_t ms;
    mach_msg_type_number_t memsz, vmsz;

    /* parse our command line options */
    while ((c = getopt(argc, argv, "bkmgs:Vh?")) != -1) {
        if (c == 'b') {
            set_units(&units, BYTES);
        } else if (c == 'k') {
            set_units(&units, KILOBYTES);
        } else if (c == 'm') {
            set_units(&units, MEGABYTES);
        } else if (c == 'g') {
            set_units(&units, GIGABYTES);
        } else if (c == 's') {
            poll = atoi(optarg);
        } else if (c == 'V') {
            printf("darwin-free version %s\n", _FREE_VERSION);
            return EXIT_SUCCESS;
        } else {
            printf(FREE_USAGE, basename(argv[0]));

            if (c == 'h' || c == '?') {
                return EXIT_SUCCESS;
            }

            return EXIT_FAILURE;
        }
    }

    if (units == -1) {
        units = KILOBYTES;
    }

    host = mach_host_self();
    task = mach_task_self();

    /* set some preferred sizes */
    memsz = sizeof(hbi) / sizeof(integer_t);
    vmsz = sizeof(hs) / sizeof(integer_t);

    /* get basic system information */
    ke = host_info(host, HOST_BASIC_INFO, (host_info_t) &hbi, &memsz);
    if (ke != KERN_SUCCESS) {
        mach_error("host_info", ke);
        return EXIT_FAILURE;
    }

    /* loop over this in case we are polling */
    while (1) {
        /* get the virtual memory page size for this machine */
        if ((ke = host_page_size(host, &hps)) != KERN_SUCCESS) {
            mach_error("host_page_size", ke);
            return EXIT_FAILURE;
        }

        /* gather virtual memory statistics */
        ke = host_statistics(host, HOST_VM_INFO, (host_info_t) &hs, &vmsz);
        if (ke != KERN_SUCCESS) {
            mach_error("host_statistics", ke);
            return EXIT_FAILURE;
        }

        /* select divisor */
        if (units == KILOBYTES) {
            div = 1024;
        } else if (units == MEGABYTES) {
            div = 1048576;    // 1024 ^ 2
        } else if (units == GIGABYTES) {
            div = 1073741824; // 1024 ^ 3
        }

        /* we have collected data, put it into our structure */
        ms.total = hbi.max_mem / div;
        ms.used = ((hs.active_count + hs.inactive_count + hs.wire_count)
                   * hps) / div;
        ms.free = (hs.free_count * hps) / div;
        ms.active = (hs.active_count * hps) / div;
        ms.inactive = (hs.inactive_count * hps) / div;
        ms.wired = (hs.wire_count * hps) / div;

        /* display the memory usage statistics */
        printf("%18s %10s %10s %10s %10s %10s\n",
               "total", "used", "free", "active", "inactive", "wired");
        printf("Mem: %13llu %10llu %10llu %10llu %10llu %10llu\n",
               ms.total, ms.used, ms.free,
               ms.active, ms.inactive, ms.wired);

        /* does the loop continue? */
        if (poll != 0) {
            sleep(poll);
            printf("\n");
        } else {
            break;
        }
    }

    return EXIT_SUCCESS;
}
